pkg/agent/tcpconn/disruptor.go

```diff
@@ -0,0 +1,141 @@
+// Package tcpconn contains a TCP connection disruptor.
+package tcpconn
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"math/rand"
+	"time"
+
+	"github.com/florianl/go-nfqueue"
+	"github.com/grafana/xk6-disruptor/pkg/iptables"
+	"github.com/grafana/xk6-disruptor/pkg/runtime"
+)
+
+// Disruptor applies TCP Connection disruptions by dropping connections according to a Dropper. A filter decides which
+// connections are considered for dropping.
+type Disruptor struct {
+	Executor runtime.Executor
+	Dropper  Dropper
+	Filter   Filter
+}
+
+// Filter holds the matchers used to know which traffic should be intercepted.
+type Filter struct {
+	// Port is the target port to match which connections will be intercepted.
+	Port uint
+}
+
+// ErrDurationTooShort is returned when the supplied duration is smaller than 1s.
+var ErrDurationTooShort = errors.New("duration must be at least 1 second")
+
+// Apply starts the disruption by subjecting connections that match the configured Filter to the Dropper.
+func (d Disruptor) Apply(ctx context.Context, duration time.Duration) error {
+	if duration < time.Second {
+		return ErrDurationTooShort
+	}
+
+	// Set context timeout only if duration is not 0 (forever).
+	ctx, cancel := context.WithTimeout(ctx, duration)
+	defer cancel()
+
+	iptables := iptables.New(d.Executor)
+	//nolint:errcheck // Nothing to do while we don't implement logging.
+	defer iptables.Remove()
+
+	config := randomNFQConfig()
+	for _, r := range d.rules(config) {
+		err := iptables.Add(r)
+		if err != nil {
+			return err
+		}
+	}
+
+	queue, err := nfqueue.Open(&nfqueue.Config{
+		NfQueue:  config.queueID,
+		Copymode: nfqueue.NfQnlCopyPacket, // Copymode must be set to NfQnlCopyPacket to be able to read the packet.
+
+		// TODO: Refine this magic value. Larger values will cause nfqueue to error such as:
+		// netlink receive: recvmsg: no buffer space available
+		// Likely this means that we're trying to use too much memory for this queue.
+		MaxQueueLen:  32,
+		MaxPacketLen: 0xffff, // TODO: This can probably be smaller for IPv4 on top of ethernet (1500 mtu).
+	})
+	if err != nil {
+		return fmt.Errorf("creating nfqueue: %w", err)
+	}
+
+	//nolint:errcheck
+	defer queue.Close()
+
+	err = queue.RegisterWithErrorFunc(ctx,
+		func(packet nfqueue.Attribute) int {
+			if d.Dropper.Drop(*packet.Payload) {
+				_ = queue.SetVerdictWithMark(*packet.PacketID, nfqueue.NfRepeat, int(config.rejectMark))
+				return 0
+			}
+
+			_ = queue.SetVerdict(*packet.PacketID, nfqueue.NfAccept)
+
+			return 0
+		},
+		func(err error) int {
+			// TODO: Handle errors.
```

Comment:
I think we should return a non `0` value and return an error from the disruptor to signal this error and facilitate debugging.

Reply:
Makes sense, I've implemented that here: https://github.com/grafana/xk6-disruptor/pull/357/commits/49db6df25821fe23400a2a052c548113096194da

I've manually testing this by hardcoding a very large `MaxQueueLen`, that we know to cause the nfqueue to fail. Upon receiving the first packet, the agent process now exits and reports the error, cleaning rules successfully and eventually exiting with rc 1.

```
root@Archiroo
11:45:41 ~ #> /home/roobre/Devel/xk6-disruptor/images/agent/build/xk6-disruptor-agent-linux-amd64 tcp-drop --rate 0 -d 9999h -p 6666
reading packet from NFQueue: netlink receive: recvmsg: no buffer space available

root@Archiroo
11:47:20 ~ #> echo $?
1
```

We should keep in mind that this might make disruptions fail mid-test in non-fatal conditions, although they will give us more visibility on errors.